### PR TITLE
[WIP] Feature: Keyboard Shortcuts (rebased on master)

### PR DESCRIPTION
**Pull request by seveibar**
Original work by @puskuruk but for some reason github stopped tracking authorship after rebase

### CODE REVIEW

File: src/Annotator/index.js [hunk 1]
```diff
@@ -102,19 +104,30 @@ export const Annotator = ({
...
-      dispatchToReducer(action)
+    if (action.type === "HEADER_BUTTON_CLICKED") {
+      if (["Exit", "Done", "Save", "Complete"].includes(action.buttonName)) {
+        return onExit(without(state, "history"))
+      } else if (action.buttonName === "Next") {
+        return onNextImage(without(state, "history"))
+      } else if (action.buttonName === "Prev") {
+        return onPrevImage(without(state, "history"))
+      }
     }
```
Comment (seveibar) [inline src/Annotator/index.js]: intercept action.type === "NEXT_IMAGE"  and call onNextImage if it's defined. Same for onPrevImage

condition should be changed to `if (action.buttonName === "Next" && onNextImage) {`

File: src/DemoSite/Editor.js [hunk 2]
```diff
@@ -123,12 +123,13 @@ const Editor = ({ onOpenAnnotator, lastOutput }: any) => {
...
-                selectedExample === "Custom"
-                  ? loadSavedInput()
-                  : examples[selectedExample]
-              )
+            onClick={() =>{
+                onOpenAnnotator(
+                  selectedExample === "Custom"
+                    ? loadSavedInput()
+                    : examples[selectedExample]
+                )
```
Comment (seveibar) [inline src/DemoSite/Editor.js]: this should be run through prettier

File: src/MainLayout/types.js [hunk 3]
```diff
@@ -127,5 +127,5 @@ export type Action =
   | {| type: "CLOSE_REGION_EDITOR", region: Region |}
   | {| type: "DELETE_REGION", region: Region |}
   | {| type: "HEADER_BUTTON_CLICKED", buttonName: string |}
-  | {| type: "SELECT_TOOL", selectedTool: string |}
+  | {| type: "SELECT_TOOL", selectedTool: ToolEnum |}
   | {| type: "CANCEL" |}
```
Comment (seveibar) [inline src/MainLayout/types.js]: need to add new actions here...

`{| type: "GO_TO_NEXT_IMAGE" |}`

`{| type: "GO_TO_PREV_IMAGE" |}`

File: src/Shortcuts/index.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  "create-pixel": {
+    name: "Create Pixel"
+  },
+  "prev-image": {
+    //TODO: { type: "GO_TO_PREV_IMAGE" }
+    name: "Previous Image"
+  },
+  "next-image": {
+    //TODO: { type: "GO_TO_NEXT_IMAGE" }
+    name: "Next Image"
```
Comment (seveibar) [inline src/Shortcuts/index.js]: each of these should have static actions defined. e.g.

```
// ...
"next-image": {
  "name": "Next Image",
  "action": { type: "GO_TO_NEXT_IMAGE"  }
}
// ...
```

File: src/Shortcuts/index.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      for (const actionId in shortcuts) {
+        const shortcut = shortcuts[actionId]
+        if (!shortcut || !shortcut.key) {
+          continue
+        }
+        if (e.key === shortcut.key) {
+          onShortcutActionDispatched({
+            type: "SELECT_TOOL",
+            selectedTool: actionId
+          })
```
Comment (seveibar) [inline src/Shortcuts/index.js]: change this to...

```javascript
onShortcutActionDispatched(shortcut.action)
```

**Comment by seveibar** [comment]
In `image-reducer.js` you'll want to grab the action for `GO_TO_NEXT_IMAGE` and `GO_TO_PREV_IMAGE` and change the state accordingly. It'll look something like this...

```
//...
case "GO_TO_NEXT_IMAGE":
  return setIn(state, ["selectedImage"], (currentImageIndex + 1) % state.images.length)

//...
```

File: src/DemoSite/Editor.js (hunk 2 above)
Comment (puskuruk) [reply src/DemoSite/Editor.js]: #34 9fd6df7577886b2511c947c04aa0eb11126c56a8

**Pull request merged**